Follow-up on `[Embedded]` — I actually tried it (mark all ~24 contract types `[Embedded]` + add the `EmbeddedAttribute` polyfill to the standalone consumer), and it **breaks the build**, which nicely pinpoints the real constraint:

```
VoidResponseSerializer.cs(12,31): error CS0103: The name 'VoidResponseFieldsId' does not exist in the current context
  [Microsoft.Testing.Extensions.MSBuild / HangDump / Retry / TrxReport / ...]
```

Why: several extensions compile `IPC/Serializers/VoidResponseSerializer.cs` **as shared source** (it's itself `[Embedded]`), but they do **not** compile `ObjectFieldIds.cs` — they resolve `VoidResponseFieldsId` from `Microsoft.Testing.Platform` via **`InternalsVisibleTo`**. Marking `VoidResponseFieldsId` (and siblings) `[Embedded]` hides them across the assembly boundary, so every extension that source-shares `VoidResponseSerializer` can no longer see the field-id → `CS0103`.

So today the design is inconsistent: the *serializer* is source-shared + `[Embedded]`, but the *field-ids it uses* are IVT-shared plain `internal`. To make the field-ids/constants `[Embedded]`, they'd also have to be **source-shared** into every assembly that uses them (HangDump, Retry, TrxReport, Extensions.MSBuild, …), not handed out via IVT — i.e. a broader restructure of how these IPC bits are shared, well beyond this contract-package PR.

**Proposal:** keep the contract types plain `internal` in this PR (they're non-public in any consumer regardless), and tackle `[Embedded]` as part of a dedicated change that converts the field-id/constants sharing from IVT to source across the platform + extensions (ideally together with the models/serializers follow-up). I've reverted the `[Embedded]` experiment. WDYT?

--- a/src/Platform/Microsoft.Testing.Extensions.HangDump/Microsoft.Testing.Extensions.HangDump.csproj
+++ b/src/Platform/Microsoft.Testing.Extensions.HangDump/Microsoft.Testing.Extensions.HangDump.csproj
@@ -33,6 +33,7 @@ This package extends Microsoft Testing Platform to provide an implementation of 
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\NamedPipeSerializer.cs" Link="IPC\Serializers\NamedPipeSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\VoidResponseSerializer.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Models\VoidResponse.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\ObjectFieldIds.cs" Link="IPC\ObjectFieldIds.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\Resources\PlatformResources.cs" Link="Resources\PlatformResources.cs" />
 
     <!-- Embedded helpers from Microsoft.Testing.Platform -->

--- a/src/Platform/Microsoft.Testing.Extensions.MSBuild/Microsoft.Testing.Extensions.MSBuild.csproj
+++ b/src/Platform/Microsoft.Testing.Extensions.MSBuild/Microsoft.Testing.Extensions.MSBuild.csproj
@@ -14,6 +14,7 @@
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\NamedPipeSerializer.cs" Link="IPC\Serializers\NamedPipeSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\VoidResponseSerializer.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Models\VoidResponse.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\ObjectFieldIds.cs" Link="IPC\ObjectFieldIds.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\Resources\PlatformResources.cs" Link="Resources\PlatformResources.cs" />
 
     <!-- Embedded helpers from Microsoft.Testing.Platform -->

--- a/src/Platform/Microsoft.Testing.Extensions.Retry/Microsoft.Testing.Extensions.Retry.csproj
+++ b/src/Platform/Microsoft.Testing.Extensions.Retry/Microsoft.Testing.Extensions.Retry.csproj
@@ -16,6 +16,7 @@
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\NamedPipeSerializer.cs" Link="IPC\Serializers\NamedPipeSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\VoidResponseSerializer.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Models\VoidResponse.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\ObjectFieldIds.cs" Link="IPC\ObjectFieldIds.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\Resources\PlatformResources.cs" Link="Resources\PlatformResources.cs" />
 
     <!-- Embedded helpers from Microsoft.Testing.Platform -->

--- a/src/Platform/Microsoft.Testing.Extensions.TrxReport/Microsoft.Testing.Extensions.TrxReport.csproj
+++ b/src/Platform/Microsoft.Testing.Extensions.TrxReport/Microsoft.Testing.Extensions.TrxReport.csproj
@@ -16,6 +16,7 @@
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\NamedPipeSerializer.cs" Link="IPC\Serializers\NamedPipeSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\VoidResponseSerializer.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Models\VoidResponse.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\ObjectFieldIds.cs" Link="IPC\ObjectFieldIds.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\Resources\PlatformResources.cs" Link="Resources\PlatformResources.cs" />
 
     <!-- Embedded helpers from Microsoft.Testing.Platform -->

--- a/src/Platform/Microsoft.Testing.Platform/ServerMode/DotnetTest/IPC/Constants.cs
+++ b/src/Platform/Microsoft.Testing.Platform/ServerMode/DotnetTest/IPC/Constants.cs
@@ -1,8 +1,11 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.CodeAnalysis;
+
 namespace Microsoft.Testing.Platform.IPC;
 
+[Embedded]
 internal static class TestStates
 {
     internal const byte Discovered = 0;
@@ -15,12 +18,14 @@ internal static class TestStates
     internal const byte InProgress = 7;
 }
 
+[Embedded]
 internal static class SessionEventTypes
 {
     internal const byte TestSessionStart = 0;
     internal const byte TestSessionEnd = 1;
 }
 
+[Embedded]
 internal static class HandshakeMessagePropertyNames
 {
     internal const byte PID = 0;
@@ -47,6 +52,7 @@ internal static class HandshakeMessagePropertyNames
     internal const byte OrchestratorFeature = 11;
 }
 
+[Embedded]
 internal static class HandshakeMessageHostTypes
 {
     // A regular (console or server) test host that actually runs tests.
@@ -63,6 +69,7 @@ internal static class HandshakeMessageHostTypes
     internal const string TestHostOrchestrator = "TestHostOrchestrator";
 }
 
+[Embedded]
 internal static class HandshakeMessageExecutionModes
 {
     // Standard test run.
@@ -75,6 +82,7 @@ internal static class HandshakeMessageExecutionModes
     internal const string Discover = "discover";
 }
 
+[Embedded]
 internal static class ProtocolConstants
 {
     internal const string Version = "1.0.0";

--- a/src/Platform/Microsoft.Testing.Platform/ServerMode/DotnetTest/IPC/ObjectFieldIds.cs
+++ b/src/Platform/Microsoft.Testing.Platform/ServerMode/DotnetTest/IPC/ObjectFieldIds.cs
@@ -1,27 +1,33 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using Microsoft.CodeAnalysis;
+
 namespace Microsoft.Testing.Platform.IPC;
 
 // WARNING: Please note this file needs to be kept aligned with the one in the dotnet sdk.
 // The protocol follows the concept of optional properties.
 // The id is used to identify the property in the stream and it will be skipped if it's not recognized.
 // We can add new properties with new ids, but we CANNOT change the existing ids (to support backwards compatibility).
+[Embedded]
 internal static class VoidResponseFieldsId
 {
     public const int MessagesSerializerId = 0;
 }
 
+[Embedded]
 internal static class TestHostCompletedRequestFieldsId
 {
     public const int MessagesSerializerId = 1;
 }
 
+[Embedded]
 internal static class TestHostProcessPIDRequestFieldsId
 {
     public const int MessagesSerializerId = 2;
 }
 
+[Embedded]
 internal static class CommandLineOptionMessagesFieldsId
 {
     public const int MessagesSerializerId = 3;
@@ -30,6 +36,7 @@ internal static class CommandLineOptionMessagesFieldsId
     public const ushort CommandLineOptionMessageList = 2;
 }
 
+[Embedded]
 internal static class CommandLineOptionMessageFieldsId
 {
     public const ushort Name = 1;
@@ -39,6 +46,7 @@ internal static class CommandLineOptionMessageFieldsId
     // Reserved: field ID 5 was ObsolescenceMessage, removed as unused.
 }
 
+[Embedded]
 internal static class DiscoveredTestMessagesFieldsId
 {
     public const int MessagesSerializerId = 5;
@@ -48,6 +56,7 @@ internal static class DiscoveredTestMessagesFieldsId
     public const ushort DiscoveredTestMessageList = 3;
 }
 
+[Embedded]
 internal static class DiscoveredTestMessageFieldsId
 {
     public const ushort Uid = 1;
@@ -61,12 +70,14 @@ internal static class DiscoveredTestMessageFieldsId
     public const ushort ParameterTypeFullNames = 9;
 }
 
+[Embedded]
 internal static class TraitMessageFieldsId
 {
     public const ushort Key = 1;
     public const ushort Value = 2;
 }
 
+[Embedded]
 internal static class TestResultMessagesFieldsId
 {
     public const int MessagesSerializerId = 6;
@@ -77,6 +88,7 @@ internal static class TestResultMessagesFieldsId
     public const ushort FailedTestMessageList = 4;
 }
 
+[Embedded]
 internal static class SuccessfulTestResultMessageFieldsId
 {
     public const ushort Uid = 1;
@@ -89,6 +101,7 @@ internal static class SuccessfulTestResultMessageFieldsId
     public const ushort SessionUid = 8;
 }
 
+[Embedded]
 internal static class FailedTestResultMessageFieldsId
 {
     public const ushort Uid = 1;
@@ -102,6 +115,7 @@ internal static class FailedTestResultMessageFieldsId
     public const ushort SessionUid = 9;
 }
 
+[Embedded]
 internal static class ExceptionMessageFieldsId
 {
     public const ushort ErrorMessage = 1;
@@ -109,6 +123,7 @@ internal static class ExceptionMessageFieldsId
     public const ushort StackTrace = 3;
 }
 
+[Embedded]
 internal static class FileArtifactMessagesFieldsId
 {
     public const int MessagesSerializerId = 7;
@@ -118,6 +133,7 @@ internal static class FileArtifactMessagesFieldsId
     public const ushort FileArtifactMessageList = 3;
 }
 
+[Embedded]
 internal static class FileArtifactMessageFieldsId
 {
     public const ushort FullPath = 1;
@@ -128,6 +144,7 @@ internal static class FileArtifactMessageFieldsId
     public const ushort SessionUid = 6;
 }
 
+[Embedded]
 internal static class TestSessionEventFieldsId
 {
     public const int MessagesSerializerId = 8;
@@ -137,11 +154,13 @@ internal static class TestSessionEventFieldsId
     public const ushort ExecutionId = 3;
 }
 
+[Embedded]
 internal static class HandshakeMessageFieldsId
 {
     public const int MessagesSerializerId = 9;
 }
 
+[Embedded]
 internal static class TestInProgressMessagesFieldsId
 {
     public const int MessagesSerializerId = 10;
@@ -151,6 +170,7 @@ internal static class TestInProgressMessagesFieldsId
     public const ushort TestInProgressMessageList = 3;
 }
 
+[Embedded]
 internal static class TestInProgressMessageFieldsId
 {
     public const ushort Uid = 1;

--- a/test/UnitTests/Microsoft.Testing.Platform.DotnetTestProtocolContract.UnitTests/Microsoft.Testing.Platform.DotnetTestProtocolContract.UnitTests.csproj
+++ b/test/UnitTests/Microsoft.Testing.Platform.DotnetTestProtocolContract.UnitTests/Microsoft.Testing.Platform.DotnetTestProtocolContract.UnitTests.csproj
@@ -12,6 +12,16 @@
   </ItemGroup>
 
   <!--
+    [Embedded] on the shared contract types resolves to Microsoft.CodeAnalysis.EmbeddedAttribute (not a framework
+    type). This independent consumer must provide it - the platform/extensions get it from src/Polyfills, and
+    dotnet/sdk already defines it for its copied IPC transport. We include only the polyfill (not via a package
+    reference) to mirror how an external consumer would supply the attribute.
+  -->
+  <ItemGroup>
+    <Compile Include="$(RepoRoot)src\Polyfills\EmbeddedAttribute.cs" Link="Polyfills\EmbeddedAttribute.cs" />
+  </ItemGroup>
+
+  <!--
     This project intentionally does NOT reference Microsoft.Testing.Platform's protocol types directly. It compiles
     the shared 'dotnet test' wire-contract source (ObjectFieldIds + Constants) into its own assembly via the shared
     .props, proving that the contract can live in testfx and be consumed by an independent assembly (e.g. dotnet/sdk)

--- a/test/UnitTests/Microsoft.Testing.Platform.UnitTests/Microsoft.Testing.Platform.UnitTests.csproj
+++ b/test/UnitTests/Microsoft.Testing.Platform.UnitTests/Microsoft.Testing.Platform.UnitTests.csproj
@@ -23,6 +23,8 @@
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\NamedPipeSerializer.cs" Link="IPC\Serializers\NamedPipeSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Serializers\VoidResponseSerializer.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\IPC\Models\VoidResponse.cs" Link="IPC\Serializers\VoidResponseSerializer.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\ObjectFieldIds.cs" Link="IPC\ObjectFieldIds.cs" />
+    <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\ServerMode\DotnetTest\IPC\Constants.cs" Link="IPC\Constants.cs" />
     <Compile Include="$(RepoRoot)src\Platform\Microsoft.Testing.Platform\Resources\PlatformResources.cs" Link="Resources\PlatformResources.cs" />
 
     <!-- Embedded helpers from Microsoft.Testing.Platform -->

